src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_consumer_power_formula.py

```diff
@@ -50,95 +50,141 @@ def generate(self) -> FormulaEngine[Power]:
-                builder.push_component_metric(
-                    successor.component_id, nones_are_zeros=nones_are_zeros
-                )
+            builder.push_component_metric(
+                grid_meter.component_id, nones_are_zeros=False
+            )
+
+        # push all non consumer components and subtract them from the grid meters
+        for component in non_consumer_components:
+            builder.push_oper("-")
+            builder.push_component_metric(
+                component.component_id,
+                nones_are_zeros=component.category != ComponentCategory.METER,
+            )
 
         return builder.build()
 
     def _gen_without_grid_meter(
         self,
         builder: ResampledFormulaBuilder[Power],
-        grid_successors: abc.Iterable[Component],
+        grid: Component,
     ) -> FormulaEngine[Power]:
         """Generate formula for calculating consumer power without a grid meter.
 
         Args:
             builder: The formula engine builder.
-            grid_successors: The grid successors.
+            grid: The grid successors.
 
         Returns:
             A formula engine that will calculate the consumer power.
         """
-        component_graph = connection_manager.get().component_graph
-        is_first = True
-        for successor in grid_successors:
+
+        def consumer_component(component: Component) -> bool:
+            """
+            Check if a component is a consumer component.
+
+            Args:
+                component: The component to check.
+
+            Returns:
+                True if the component is a consumer component, False otherwise.
+            """
             # If the component graph supports additional types of grid successors in the
             # future, additional checks need to be added here.
-            if (
-                component_graph.is_battery_chain(successor)
-                or component_graph.is_chp_chain(successor)
-                or component_graph.is_pv_chain(successor)
-                or component_graph.is_ev_charger_chain(successor)
-            ):
-                continue
-            if not is_first:
-                builder.push_oper("+")
-            is_first = False
-            builder.push_component_metric(successor.component_id, nones_are_zeros=False)
+            return (
+                component.category
+                in {ComponentCategory.METER, ComponentCategory.INVERTER}
+                and not component_graph.is_battery_chain(component)
+                and not component_graph.is_chp_chain(component)
+                and not component_graph.is_pv_chain(component)
+                and not component_graph.is_ev_charger_chain(component)
+            )
+
+        component_graph = connection_manager.get().component_graph
+        consumer_components = component_graph.dfs(grid, set(), consumer_component)
+        print("consumer_components")
+        print(consumer_components)
```

Comment:
These `print` above look like debugging leftovers and you might want  to remove them